service/storage_service.cc

```diff
@@ -2069,6 +2077,68 @@ class topology_coordinator {
         }
     };
 
+    // Returns `true` if the node was accepted, `false` otherwise
+    future<bool> finish_join_node_handshake(node_to_work_on& node) {
+        auto ip = _address_map.find(node.id);
+        if (!ip) {
+            slogger.warn("raft topology: cannot send JOIN_NODE_RESPONSE to {} because mapping to ip is not available",
+                            node.id);
+            co_await coroutine::exception(std::make_exception_ptr(
+                    std::runtime_error(::format("no ip address mapping for {}", node.id))));
+        }
+
+        auto decision = validate_joining_node(node);
+        const bool node_accepted = std::holds_alternative<join_node_response_params::accepted>(decision);
+
+        release_guard(std::move(node.guard));
+
+        if (node_accepted && !_raft.get_configuration().contains(node.id)) {
+            co_await _raft.modify_config({raft::config_member({node.id, {}}, {})}, {});
```

Comment:
@kbr-scylla so it looks like the code is already working as we discussed today on the call. When we add the node here to group0 it is already in the topology and is processed by the topology coordinator.

Reply:
Hm, yeah it is in `none` state, which is more like being a candidate for a bootstrapping node, not yet a really bootstrapping node.

I was thinking that we should confirm that this node can bootstrap and transition its state to `bootstrapping` or something before we add it to group 0. I.e. we accept it by doing a state machine transition, and only then add it to group 0.

In particular we'd start taking the node into account when calculating features or generally making cluster-wide decisions.

Reply:
But dunno if it makes any difference in practice, whether it would help failure recovery and so on.

Reply:
No, you are correct. We do need to move those lines a little bit below after moving to the bootstrapping state. Theoretically it is possible that a coordinator will crash and choose another node to work on.

Reply:
Currently, we transition from `node_state::none` -> `node_state::bootstrapping` and also transition the topology to `commit_cdc_generation` in the same step. I'll have to separate them so that validation + transition too bootstrap is one step, and cdc generation broadcast + moving to `commit_cdc_generation` is another. Similar thing needs to be done with `replace`.

I'll also add a new state to transition the node to when we reject, so that we know we still have to deliver the rejection before definitely moving it to the `left` state.

Reply:
Why should we necessary split `commit_cdc_generation` transition?

Reply:
Otherwise it would result in weird interleaving with some of the CDC logic.

For reference, this is how we currently handle joining nodes in `none` state:

https://github.com/scylladb/scylladb/blob/bb7121a1fbdd6b3a6be76486e44f233b91a6e2d2/service/storage_service.cc#L1971-L1994

Verification of the parameters would have to be done here, then `prepare_and_broadcast_cdc_generation_data` would be called and ultimately we would change the state to `bootstrap`/`commit_cdc_generation`. Next time the topology coordinator runs, it handles the `commit_cdc_generation` state ([here](https://github.com/scylladb/scylladb/blob/bb7121a1fbdd6b3a6be76486e44f233b91a6e2d2/service/storage_service.cc#L1715)) - I suppose we could send the join_node_response there, but it just looks out of place, and would have to be skipped if `commit_cdc_generation` was issued by the user via the `check_and_repair_cdc_streams` command.

Reply:
> But it looks like making modify_config() call part of handling of commit_cdc_generation will achieve the same.

I tried this approach but it looked bad to me. Before I rebased (now, tablets added some more topology transitions) the transitions actually looked like this:

- Bootstrap requests are transitioned to `node_state::bootstrapping` + `commit_cdc_generation`
- Replacing requests are transitioned to `node_state::replacing` + `write_both_read_old`

This means that we would have to take care of adding the node to group 0 during handlers for both `commit_cdc_generation` and `write_both_read_old`; what made it worse is that `commit_cdc_generation` transitions to `write_both_read_old`. We could detect whether we need to send join_node_response with strategically placed ifs but it makes the state machine flow more convoluted than it needs in my opinion.

Reply:
OK. Indeed since bootstrap and replace goes through different states not introducing a common state makes things more complicated.